builtin/providers/google/resource_container_cluster.go

```diff
@@ -602,27 +601,127 @@ func resourceContainerClusterUpdate(d *schema.ResourceData, meta interface{}) er
 
-	req := &container.UpdateClusterRequest{
-		Update: &container.ClusterUpdate{
-			DesiredNodeVersion: desiredNodeVersion,
-		},
-	}
-	op, err := config.clientContainer.Projects.Zones.Clusters.Update(
-		project, zoneName, clusterName, req).Do()
-	if err != nil {
-		return err
-	}
+	d.Partial(true)
 
-	// Wait until it's updated
-	waitErr := containerOperationWait(config, op, project, zoneName, "updating GKE cluster", 10, 2)
-	if waitErr != nil {
-		return waitErr
+	if d.HasChange("node_pool") {
+		from, to := d.GetChange("node_pool")
+
+		// node pool configs keyed by name
+		fromMap := make(map[string]map[string]interface{})
+		toMap := make(map[string]map[string]interface{})
+
+		// node pools to add and remove
+		addNp := make([]*container.NodePool, 0, 0)
+		rmNp := make([]string, 0, 0)
+
+		for _, v := range from.([]interface{}) {
+			np := v.(map[string]interface{})
+			nameKey := np["name"].(string)
+			fromMap[nameKey] = np
+		}
+
+		for _, v := range to.([]interface{}) {
+			np := v.(map[string]interface{})
+			nameKey := np["name"].(string)
+			toMap[nameKey] = np
+
+			f, ok := fromMap[nameKey]
+			if !ok {
+				// The node pool is in 'to' but not 'from', so it is being added.
+				addNp = append(addNp, buildNodePool(np))
+			} else if np["initial_node_count"] != f["initial_node_count"] {
+				return fmt.Errorf("Cannot change property initial_node_count on node_pool %s", nameKey)
+			}
+		}
+
+		for name, _ := range fromMap {
+			_, ok := toMap[name]
+			if !ok {
+				// The node pool is in 'from' but not 'to', so it is being removed.
+				rmNp = append(rmNp, name)
+			}
+		}
+
+		for _, np := range addNp {
+			req := &container.CreateNodePoolRequest{
+				NodePool: np,
+			}
+			op, err := config.clientContainer.Projects.Zones.Clusters.NodePools.Create(project, zoneName, clusterName, req).Do()
+			if err != nil {
+				return err
+			}
+			// Wait until it's updated
+			waitErr := containerOperationWait(config, op, project, zoneName, "adding GKE node pool", 10, 2)
+			if waitErr != nil {
+				return waitErr
+			}
+
+			log.Printf("[INFO] GKE cluster %s has added a node pool %s", d.Id(), np.Name)
+		}
+
+		for _, npName := range rmNp {
+			op, err := config.clientContainer.Projects.Zones.Clusters.NodePools.Delete(project, zoneName, clusterName, npName).Do()
+			if err != nil {
+				return err
+			}
+			// Wait until it's updated
+			waitErr := containerOperationWait(config, op, project, zoneName, "deleting GKE node pool", 10, 2)
+			if waitErr != nil {
+				return waitErr
+			}
+
+			log.Printf("[INFO] GKE cluster %s has deleted a node pool %s", d.Id(), npName)
+		}
+		d.SetPartial("node_pool")
 	}
 
-	log.Printf("[INFO] GKE cluster %s has been updated to %s", d.Id(),
-		desiredNodeVersion)
+	if d.HasChange("node_version") {
+		desiredNodeVersion := d.Get("node_version").(string)
+
+		nodePoolsCount := d.Get("node_pool.#").(int)
+		reqs := make([]*container.UpdateClusterRequest, 0, 0)
```

Comment:
Not a big deal, but this could be `reqs := []*container.UpdateClusterRequest{}`.